[CIR] Upstream convert to mask builtins in CIR codegen

MarwanTarik opened the pull request:
This PR is part of https://github.com/llvm/llvm-project/issues/167752. It upstreams the codegen and tests for the convert to mask builtins implemented in the incubator, including:

Upstream X86 mask conversion builtins from clangir:
- cvtmask2b/w/d/q*
- cvtb/w/d/q2mask* 

Upstreamed helpers:
- emitX86MaskedCompare()
- emitX86ConvertToMask()
- emitX86SExtMask()

MarwanTarik committed:
Upstream CIR Codgen for convert to mask X86 builtins

MarwanTarik committed:
Merge branch 'main' into main

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -278,6 +278,113 @@ static mlir::Value emitX86MaskTest(CIRGenBuilderTy &builder, mlir::Location loc,
+                                              mlir::Value maskIn,
+                                              mlir::Location loc) {
+  if (maskIn) {
+    llvm_unreachable("NYI");
```
Why not just implement this? It's only a few lines of code.

andykaylor replied:
```suggestion
    cgf.cgm.errorNYI(loc, "emitX86MaskedCompareResult");
```

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -278,6 +278,113 @@ static mlir::Value emitX86MaskTest(CIRGenBuilderTy &builder, mlir::Location loc,
   return emitIntrinsicCallOp(builder, loc, intrinsicName, resTy,
                              mlir::ValueRange{lhsVec, rhsVec});
 }
+static mlir::Value emitX86MaskedCompareResult(CIRGenFunction &cgf,
```
Can you add a `CIRGenBuilderTy &builder` to all of these helper functions and a comment saying that the `cgf` parameter should be removed when all the NYI cases are implemented? Then replace `cgf.getBuilder()` with `builder` everywhere.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -686,6 +794,8 @@ mlir::Value CIRGenFunction::emitX86BuiltinExpr(unsigned builtinID,
   case X86::BI__builtin_ia32_cvtmask2q256:
   case X86::BI__builtin_ia32_cvtmask2q512:
+    return emitX86SExtMask(*this, ops[0], convertType(expr->getType()),
+                           getLoc(expr->getExprLoc()));
```
```suggestion
    return emitX86SExtMask(*this, builder, ops[0], convertType(expr->getType()),
                           getLoc(expr->getExprLoc()));
```

MarwanTarik committed:
Refactor x86 masked compare functions to use CIRGenBuilderTy and improve error handling

MarwanTarik committed:
Merge branch 'llvm:main' into main

MarwanTarik committed:
[CIR] [X86] Fix x86 builtin expression handling by returning mlir::Value in error cases

MarwanTarik committed:
[CodeGen][X86] Add test for _mm512_movm_epi32 intrinsic with CIR and LLVM checks

MarwanTarik commented:
I've resolved the issues. Please review the updated changes.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -278,6 +284,127 @@ static mlir::Value emitX86MaskTest(CIRGenBuilderTy &builder, mlir::Location loc,
+                                              mlir::Value maskIn,
+                                              mlir::Location loc) {
+  if (maskIn) {
+    cgf.cgm.errorNYI(loc, "emitX86MaskedCompareResult");
```
```suggestion
    cgf.cgm.errorNYI(loc, "emitX86MaskedCompareResult");
    return {};
```
Returning a null `mlir::Value` rather than `std::nullopt` indicates that we already reported an error.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -278,6 +284,127 @@ static mlir::Value emitX86MaskTest(CIRGenBuilderTy &builder, mlir::Location loc,
+  mlir::Value cmp;
+
+  if (cc == 3) {
+    cgf.cgm.errorNYI(loc, "emitX86MaskedCompare: cc == 3");
```
```suggestion
    cgf.cgm.errorNYI(loc, "emitX86MaskedCompare: cc == 3");
    return {};
```

MarwanTarik committed:
Add tests for AVX512 mask and move  intrinsics in CIR

MarwanTarik committed:
[CIR] [X86] Improve error handling in emitX86MaskedCompareResult and emitX86MaskedCompare functions

MarwanTarik commented on clang/test/CodeGen/X86/avx512vldq-builtins.c:
```diff
@@ -951,6 +951,17 @@ __m256i test_mm256_movm_epi32(__mmask8 __A) {
   return _mm256_movm_epi32(__A); 
 }
 
+__m512i test_mm512_movm_epi32(__mmask16 __A) {
```
Ahh, I’m really sorry, it’s my fault. For some reason, I got the files confused

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -278,6 +284,130 @@ static mlir::Value emitX86MaskTest(CIRGenBuilderTy &builder, mlir::Location loc,
+
+// TODO: The cgf parameter should be removed when all the NYI cases are
+// implemented.
+static std::optional<mlir::Value> emitX86SExtMask(CIRGenFunction &cgf, CIRGenBuilderTy &builder,
```
```suggestion
static std::optional<mlir::Value> emitX86SExtMask(CIRGenBuilderTy &builder,
```
The extra parameter is not needed here.

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/avx512bw-builtins.c:
```diff
@@ -768,3 +768,42 @@ unsigned char test_ktestz_mask64_u8(__mmask64 A, __mmask64 B) {
+  // CIR: %{{.*}} = cir.cast integral %{{.*}} : !cir.vector<!cir.int<s, 1> x 32> -> !cir.vector<!s16i x 32>
+  // LLVM-LABEL: @test_mm512_movm_epi16
+  // LLVM:  %{{.*}} = bitcast i32 %{{.*}} to <32 x i1>
+  // LLVM:  %{{.*}} = sext <32 x i1> %{{.*}} to <32 x i16>
```
Can you add OGCG checks here?

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/avx512dq-builtins.c:
```diff
@@ -13,6 +13,17 @@
+  // CIR: %{{.*}} = cir.cast integral %{{.*}} : !cir.vector<!cir.int<s, 1> x 8> -> !cir.vector<!s64i x 8>
+  // LLVM-LABEL: @test_mm512_movm_epi64
+  // LLVM: %{{.*}} = bitcast i8 %{{.*}} to <8 x i1>
+  // LLVM: %{{.*}} = sext <8 x i1> %{{.*}} to <8 x i64>
```
Add OGCG checks

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/avx512vldq-builtins.c:
```diff
@@ -0,0 +1,110 @@
+  // LLVM-LABEL: @test_mm_movm_epi32
+  // LLVM: %{{.*}} = bitcast i8 %{{.*}} to <8 x i1>
+  // LLVM: %{{.*}} = shufflevector <8 x i1> %{{.*}}, <8 x i1> %{{.*}}, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
+  // LLVM: %{{.*}} = sext <4 x i1> %{{.*}} to <4 x i32>
```
Add OGCG checks throughout this test

MarwanTarik committed:
Merge branch 'llvm:main' into main

MarwanTarik committed:
Remove unused 'cgf' parameter from emitX86SExtMask function

MarwanTarik committed:
Add OGCG labels for movm_epi functions in AVX512 builtins tests

MarwanTarik committed:
Remove test functions for _mm512_movm_epi8 and _mm512_movm_epi32 in AVX512 builtins tests

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/avx512bw-builtins.c:
```diff
@@ -768,3 +768,47 @@ unsigned char test_ktestz_mask64_u8(__mmask64 A, __mmask64 B) {
+
+__mmask64 test_mm512_movepi8_mask(__m512i __A) {
+  // CIR-LABEL: @_mm512_movepi8_mask
+  // CIR: %{{.*}} = cir.vec.cmp(lt, %{{.*}}, %{{.*}}) : !cir.vector<{{!s8i|!u8i}} x 64>, !cir.vector<!cir.int<u, 1> x 64>
```
This also produces a bitcast from <32 x i1> to i32. Can you add that to the checks (CIR, LLVM, and OGCG) on all the relevant cases? I realize the classic codegen equivalent of this test only checks the icmp instruction, but it should be checking the bitcast too.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -674,6 +804,10 @@ CIRGenFunction::emitX86BuiltinExpr(unsigned builtinID, const CallExpr *expr) {
+  cgm.errorNYI(expr->getSourceRange(),
+              std::string("unimplemented x86 builtin call: ") + 
+              getContext().BuiltinInfo.getName(builtinID));
+  return mlir::Value{};
```
Still

MarwanTarik committed:
Fix formatting in error handling for unimplemented x86 builtin calls

MarwanTarik committed:
Remove test cases for unsigned char in AVX512VLBW builtins

MarwanTarik committed:
Add bitcast operation to LLVM and OGCG labels for test_mm_movepi64_mask

MarwanTarik committed:
Add bitcast operations to comparison results

MarwanTarik committed:
Remove unnecessary test cases for -fno-signed-char in AVX512 builtins

MarwanTarik committed:
Add bitcast operations to comparison results in avx512dq and avx512vl builtins

MarwanTarik commented:
I added the missing `bitcast` checks and removed the `-fno-signed-char` checks from `avx512bw-builtins.c` and `avx512vlbw-builtins.c`

andykaylor reviewed (approved):
lgtm

MarwanTarik committed:
Merge branch 'main' into main

MarwanTarik committed:
Fix AVX512 tests: vector syntax, C++ compat, disable broken lowering

MarwanTarik commented:
I figured out that there were some bugs in the tests. I could summarize it in the following:
- Removed @ prefix from CIR-LABEL patterns to match C++ mangled names
- Updated OGCG-LABEL patterns from @test_function to {{.*}}test_function{{.*}}( for C/C++ compatibility
- Fixed vector syntax from old !cir.vector<!cir.int<s, 1> x 32> to new !cir.vector<32 x !cir.int<u, 1>> format 

Also, I figured out that it seems there's a backend bug:  the CIR-to-LLVM lowering incorrectly rejects valid 1-bit to N-bit integer extensions in vectors. So I disabled CIR-to-LLVM RUN lines temporarily as a workaround. Pls tell me your thoughts about that

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/avx512bw-builtins.c:
```diff
@@ -1,32 +1,21 @@
-// RUN: FileCheck --check-prefixes=LLVM --input-file=%t.ll %s
-// RUN: %clang_cc1 -x c -flax-vector-conversions=none -ffreestanding %s -triple=x86_64-unknown-linux  -target-feature +avx512bw -fno-signed-char  -fclangir -emit-llvm -o %t.ll -Wall -Werror -Wsign-conversion
-// RUN: FileCheck --check-prefixes=LLVM --input-file=%t.ll %s
+// FIXME: CIR to LLVM lowering fails with "integer width of the output type is smaller or equal to the integer width of the input type" error
```
I found the problem. It was a bad assumption in our lowering of VecCmpOp. This change should fix it.

```
diff --git a/clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp b/clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp
index 7c9cf8e2c2e2..4dcde56f2917 100644
--- a/clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp
+++ b/clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp
@@ -3605,9 +3605,13 @@ mlir::LogicalResult CIRToLLVMVecCmpOpLowering::matchAndRewrite(
   }

   // LLVM IR vector comparison returns a vector of i1. This one-bit vector
-  // must be sign-extended to the correct result type.
-  rewriter.replaceOpWithNewOp<mlir::LLVM::SExtOp>(
-      op, typeConverter->convertType(op.getType()), bitResult);
+  // must be sign-extended to the correct result type, unless a vector of i1 is
+  // the type we need.
+  if (cast<cir::IntType>(cast<cir::VectorType>(op.getType()).getElementType())
+          .getWidth() > 1)
+    rewriter.replaceOpWithNewOp<mlir::LLVM::SExtOp>(
+        op, typeConverter->convertType(op.getType()), bitResult);
+  else
+    rewriter.replaceOp(op, bitResult);
   return mlir::success();
 }


```
Can you add that to your PR?

MarwanTarik replied:
I’ve added it.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -279,6 +285,102 @@ static mlir::Value emitX86MaskTest(CIRGenBuilderTy &builder, mlir::Location loc,
+                                  mlir::Value op,
+                                  mlir::Type dstTy, mlir::Location loc) {
+  unsigned numberOfElements = cast<cir::VectorType>(dstTy).getSize();
+  mlir::Value mask = getMaskVecValue(builder, loc, op, numberOfElements);
```
https://github.com/llvm/llvm-project/pull/172912

MarwanTarik committed:
Fix sign-extension logic in vector comparison lowering

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/avx512bw-builtins.c:
```diff
@@ -1,32 +1,21 @@
-// RUN: %clang_cc1 -x c -flax-vector-conversions=none -ffreestanding %s -triple=x86_64-unknown-linux  -target-feature +avx512bw -fno-signed-char  -fclangir -emit-llvm -o %t.ll -Wall -Werror -Wsign-conversion
-// RUN: FileCheck --check-prefixes=LLVM --input-file=%t.ll %s
+// FIXME: CIR to LLVM lowering fails with "integer width of the output type is smaller or equal to the integer width of the input type" error
+// RUN-DISABLED: %clang_cc1 -x c -flax-vector-conversions=none -ffreestanding %s -triple=x86_64-unknown-linux -target-feature +avx512bw  -fclangir -emit-llvm -o %t.ll -Wall -Werror -Wsign-conversion
```
Can you re-enable these runs now?

MarwanTarik replied:
ok

MarwanTarik committed:
Enable LLVM lowering for AVX512 builtins tests

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/avx512bw-builtins.c:
```diff
@@ -768,3 +756,51 @@ unsigned char test_ktestz_mask64_u8(__mmask64 A, __mmask64 B) {
+  
+  // LLVM-LABEL: @test_mm512_movm_epi16
+  // LLVM:  %{{.*}} = bitcast i32 %{{.*}} to <32 x i1>
+  // LLVM:  %{{.*}} = sext <32 x i1> %{{.*}} to <32 x i16>
```
I think this test, among others, will fail because of the sign-extend issue. I'll wait until I've merged that before starting the test run for this PR. Hopefully you won't have to make any changes.

andykaylor replied:
Oh, I was wrong. It looks like you'll need to rebase and update the checks in this PR that are looking for `<N x !cir.int<u, 1>>`

MarwanTarik committed:
Merge branch 'main' into main

MarwanTarik committed:
Merge branch 'llvm:main' into main

MarwanTarik committed:
Update CIR casts in AVX512 builtins to use signed integer types

MarwanTarik committed:
Fix LLVM labels in AVX512 builtins test cases

MarwanTarik commented:
I think tests should pass now

MarwanTarik commented:
@andykaylor

andykaylor commented:
@MarwanTarik Everything looks great except that the code formatting check is failing. You need to run clang-format on your changes. THis is the command that I use:

`git diff -U0 --no-color --relative origin/HEAD^ | <build-install-dir>/clang-format-diff.py -p1 -i`

That will format your files in place, and you can just commit the changes and push to your branch.

MarwanTarik committed:
Refactor formatting

MarwanTarik commented:
I fixed the formatting, and I really want to thank you, Andy, for your help and guidance through this pull request.
